src/schema/v1/filter_artworks.ts

```diff
@@ -384,13 +359,70 @@ const filterArtworksTypeFactory = (
       loader = loaderWithCache
     }
 
-    const blacklistedFields = ["artworks_connection", "__id"]
-    if (queriedForFieldsOtherThanBlacklisted(fieldNodes, blacklistedFields)) {
-      return loader(gravityOptions).then(response =>
-        Object.assign({}, response, { options: gravityOptions })
-      )
+    // Introspect connection args for `artworks_connection`.
+    // Use those and fetch all data in one request.
+    // Resolver for `artworks_connection` just creates a connection
+    // from the already fetched data, instead of re-fetching.
+
+    const connectionArgs = {}
+    info.fieldNodes &&
+      info.fieldNodes.forEach(rootNode => {
+        const visitor = (
+          fieldNode: FieldNode | FragmentDefinitionNode,
+          fragments?: { [key: string]: FragmentDefinitionNode }
+        ): void => {
+          visit(fieldNode, {
+            Field(node, _key, _parent, path, _ancestors) {
+              // Stop recursion for nodes deeper than our threshold
+              if (path.length > SELECTION_DEPTH_THRESHOLD) {
+                return
+              }
+              if (
+                path.length === SELECTION_DEPTH_THRESHOLD &&
+                node.name.value === "artworks_connection"
+              ) {
+                node.arguments &&
+                  node.arguments.forEach(arg => {
+                    if (
+                      ["first", "last", "before", "after"].includes(
+                        arg.name.value
+                      )
+                    ) {
+                      const val = (arg as any).value.value
+                        ? (arg as any).value.value
+                        : 0
+
+                      connectionArgs[arg.name.value] = parseInt(val) || val
+                    }
+                  })
+                return BREAK
+              }
+            },
+            FragmentSpread(node) {
+              const fragmentDef = info.fragments[node.name.value]
+              visitor(fragmentDef, fragments)
+            },
+          })
+        }
+
+        visitor(rootNode)
+      })
+
+    let relayOptions: any = {}
+    if (Object.keys(connectionArgs).length) {
+      relayOptions = convertConnectionArgsToGravityArgs(connectionArgs)
     }
-    return { hits: null, aggregations: null, options: gravityOptions }
+
+    if (!!gravityOptions.page) relayOptions.page = gravityOptions.page
+
+    // If only queried for __id, can just return w/o fetching.
+    if (!includesFieldsOtherThanSelectionSet(info, ["__id"])) {
+      return { hits: null, aggregations: null, options: gravityOptions }
+    }
+
+    return loader(Object.assign(gravityOptions, relayOptions, {})).then(
+      response => Object.assign({}, response, { options: gravityOptions })
```

Comment:
We probably don't need to propagate these options through in the response anymore, too.